### PR TITLE
Workshop Properties

### LINKED ISSUE #937: Sync: Workshop Properties
## Description

1. Sync the values in the **Synced** list below.
2. Create an end to end (E2E) tests for all properties (1 for server changing the value and 1 for client changing the value if applicable)
3. Create a command to test changing each property in game [see similar command](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/blob/development/source/GameInterface/Services/Towns/Commands/TownDebugCommand.cs#L455)

Also see additional information at the bottom for more information and templates

**Key**
Server side only (only allow the running of the function on the server side)
Client side only (only allow the running of the function on the client side)
Client side sync (request the server to change the value, server then changes the value and replies to the client allowing the change)
Server side sync (only server allows running of the function and send to all clients the value changed and clients update the value/call original function)

**Synced**

| Method/Field Name | Sync Type | Additional Information |
|:-----------------|:---------:|:--------------|
| Capital | Server Side Sync | - |
| Expense | Server Side Sync | Is Campaign.Current.Models.WorkshopModel.DailyExpense constant? If so, no need to sync. |
| InitialCapital | Server Side Sync | - |
| LastRunCampaignTime | Server Side Sync | - |
| WorkshopType | Server Side Sync | - |

**Deferred**

None

**Externally Deferred**

None

**Non-Synced**

None

## Intended Design

You are expected to only sync the property setters.
Try using the new [autosync tool](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/wiki/Property-AutoSync)

## Location

Create a branch based from [development](https://github.com/Bannerlord-Coop-Team/BannerlordCoop)

## Related Issues

Blocked by:
- #935

## Requirements

N/A

## Additional information

Using the [GameInterface Service Tempale](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/development/source/GameInterface/Services/Template) create commands to test the synced values.

## Definition of Done
- [ ] Class level comments exist for all new classes.
- [ ] XUnit E2E tests exist for every property on the server (and client if applicable)
- [ ] Each property has a command getter and setter

### PR DESCRIPTION
**Pull request by EgardA**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] All new classes have class-level documentation comments, if there are any at all
- [x] Tests for the changes have been added (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation update
- [ ] Other... Please describe:

## What is the current behavior?

No workshop properties

## What is the new behavior?

Resolves #937

### CODE REVIEW

**Comment by EgardA** [comment]
This still has some work, hold off for now

File: source/E2E.Tests/Services/Workshops/WorkshopPropertyTests.cs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            var workshopType = GameObjectCreator.CreateInitializedObject<WorkshopType>();
+            var workshopType2 = GameObjectCreator.CreateInitializedObject<WorkshopType>();
+
+            Server.ObjectManager.AddNewObject(workshopType, out string typeId);
+            Server.ObjectManager.AddNewObject(workshopType2, out string typeId2);
+            Server.ObjectManager.AddNewObject(workshop, out string workshopId);
+
+            WorkshopId = workshopId;
+            WorkshopTypeId = typeId;
+            WorkshopTypeId2 = typeId2;
```
Comment (garrettluskey) [inline source/E2E.Tests/Services/Workshops/WorkshopPropertyTests.cs]:
```suggestion
            Server.ObjectManager.AddNewObject(workshopType, out WorkshopTypeId);
            Server.ObjectManager.AddNewObject(workshopType2, out  WorkshopTypeId2);
            Server.ObjectManager.AddNewObject(workshop, out WorkshopId);

            WorkshopId = workshopId;
            WorkshopTypeId = typeId;
            WorkshopTypeId2 = typeId2;
```

File: source/E2E.Tests/Services/Workshops/WorkshopPropertyTests.cs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    public class WorkshopPropertyTests : IDisposable
+    {
+        E2ETestEnvironment TestEnvironement { get; }
+
+        EnvironmentInstance Server => TestEnvironement.Server;
+        IEnumerable<EnvironmentInstance> Clients => TestEnvironement.Clients;
+
+        string WorkshopId { get; set; }
+        string WorkshopTypeId { get; set; }
+        string WorkshopTypeId2 { get; set; }
```
Comment (garrettluskey) [inline source/E2E.Tests/Services/Workshops/WorkshopPropertyTests.cs]:
```suggestion
        string WorkshopId { get; }
        string WorkshopTypeId { get; }
        string WorkshopTypeId2 { get; }
```

File: source/GameInterface/Services/Workshops/Patches/WorkshopPropertyPatches.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+{
+    public enum PropertyType
+    {
+        Capital,
+        LastRunCampaignTime,
+        WorkshopType
+    }
+
+    [HarmonyPatch(typeof(Workshop))]
+    public class WorkshopPropertyPatches
```
Comment (garrettluskey) [inline source/GameInterface/Services/Workshops/Patches/WorkshopPropertyPatches.cs:24]: We need the following properties synced still.
Also for future reference could you comment in the issue why you are skipping some sync?

- InitialCapital (set in the constructor which requires sync)
- Capital (set in the constructor which requires sync)

**Review by garrettluskey — CHANGES_REQUESTED** [review]

**Comment by EgardA** [comment]
Do not merge until WorkshopType has been synced

**Review by garrettluskey — APPROVED** [review]

**Pull request merged**